### PR TITLE
Rework jemalloc extension

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR reworks the jemalloc extension to use C code, not our modified C++ code. I've also updated the version, and it should now be much easier to update to newer jemalloc versions in the future (`README.md` explains how).

Following the discussion in https://github.com/duckdb/duckdb/discussions/11455, I've changed many configuration settings in jemalloc, which should slightly improve performance/portability. We should be able to activate jemalloc on ARM64 again, but I'll leave that for a future PR.

I've also added a settings that allows our users to toggle jemalloc's background threads:
```sql
SET allocator_background_threads=true;
SET allocator_background_threads=false;
```
This setting is disabled by default. It spawns one background thread for every 32 logical CPUs (at least 1). Enabling this setting slightly improves performance further, especially on many-core machines.

I tried letting jemalloc override the C++ new/delete operators, which worked great everywhere except for Python/Julia, which I guess also override allocations, so there was some kind of conflict there. I've disabled this for now.

I plan to tweak the settings further based on `jemalloc_stats`, but I'll also leave that for a future PR.

### CODE REVIEW

**Comment by lnkuiper** [comment]
Python Arrow regression tests are failing. The rework seems to have introduced a regression there.  [In my own CI, the same regression test passed](https://github.com/lnkuiper/duckdb/actions/runs/8909944653/job/24468195924), so it seems to depend on the machine.

In my own tests (on an AWS box with 8 cores/16 threads, TPC-H), the rework only seemed to improve performance.

**Comment by Tishj** [comment]
> Python Arrow regression tests are failing. The rework seems to have introduced a regression there. [In my own CI, the same regression test passed](https://github.com/lnkuiper/duckdb/actions/runs/8909944653/job/24468195924), so it seems to depend on the machine.
> 
> In my own tests (on an AWS box with 8 cores/16 threads, TPC-H), the rework only seemed to improve performance.

That looks like it's just spurious, the python regression tests haven't been stable for a bit now

File: src/parallel/task_scheduler.cpp [hunk 1]
```diff
@@ -144,8 +142,18 @@ void TaskScheduler::ExecuteForever(atomic<bool> *marker) {
 	shared_ptr<Task> task;
 	// loop until the marker is set to false
 	while (*marker) {
-		// wait for a signal with a timeout
-		queue->semaphore.wait();
+		if (allocator_background_threads) {
```
Comment (Mytherin) [inline src/parallel/task_scheduler.cpp:145]: Maybe `|| Allocator::SupportsFlush()` so the path below is not called outside of jemalloc?

File: src/parallel/task_scheduler.cpp [hunk 2]
```diff
@@ -144,8 +142,18 @@ void TaskScheduler::ExecuteForever(atomic<bool> *marker) {
...
+			// background threads clean up allocations, just start an untimed wait
+			queue->semaphore.wait();
+		} else if (!queue->semaphore.wait(500000)) {
```
Comment (Mytherin) [inline src/parallel/task_scheduler.cpp:148]: Maybe this can be retrieved from the allocator as well?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great - some minor comments:

**Comment by lnkuiper** [comment]
Thanks for the feedback! This should be good to go once CI passes

**Comment by hannes** [comment]
Thanks!

**Pull request merged**